Engine/Signals/DefaultSignalHandler.cs

```diff
@@ -0,0 +1,259 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using Newtonsoft.Json;
+using QuantConnect.Algorithm.Framework.Signals;
+using QuantConnect.Algorithm.Framework.Signals.Analysis;
+using QuantConnect.Algorithm.Framework.Signals.Analysis.Providers;
+using QuantConnect.Interfaces;
+using QuantConnect.Logging;
+using QuantConnect.Packets;
+
+namespace QuantConnect.Lean.Engine.Signals
+{
+    /// <summary>
+    /// Base signal handler that supports sending signals to the messaging handler
+    /// </summary>
+    public class DefaultSignalHandler : ISignalHandler
+    {
+        private static readonly IReadOnlyCollection<SignalScoreType> ScoreTypes = SignalManager.ScoreTypes;
+
+        private DateTime _nextPersistenceUpdate;
+        private DateTime _nextChartUpdateAlgorithmTimeUtc;
+
+        private IMessagingHandler _messagingHandler;
+        private CancellationTokenSource _cancellationTokenSource;
+        private readonly ConcurrentQueue<Packet> _messages = new ConcurrentQueue<Packet>();
+        private readonly Dictionary<SignalScoreType, Series> _seriesByScoreType = new Dictionary<SignalScoreType, Series>();
+
+        /// <inheritdoc />
+        public bool IsActive => !_cancellationTokenSource.IsCancellationRequested;
```

Comment:
Does this need to be checked for null?

Reply:
Absolutely. It is unlikely, however, that one would check `IsActive` before calling `Run`, but I completely agree with you. Change to be available in next push.